src/dictionary.c

```diff
@@ -0,0 +1,103 @@
+
+/*
+ * A dictionary data structure based on tries
+ *
+ * See dictionary.h for function documentation.
+ */
+
+#include <stdlib.h>
+#include <stdbool.h>
+#include <assert.h>
+#include <unistd.h>
+#include "utils.h"
+#include "dictionary.h"
+
+/* See dictionary.h */
+dict_t* dict_new() {
+
+    dict_t *d;
+    int rc;
+
+    d = malloc(sizeof(dict_t));
+
+    if (d == NULL) {
+        error("Could not allocate memory");
+        return NULL;
+    }
+
+    rc = dict_init(d);
+    if (rc != 0) {
+        error("Could not initialize dict");
+        return NULL;
+    }
+
+    return d;
+}
+
+/* See dictionary.h */
+int dict_init(dict_t *d) {
+    assert(d != NULL);
+
+    trie_t *t = trie_new();
+    if (t == NULL) {
+        return 1;
```

Comment:
`stdlib.h` provides `EXIT_SUCCESS` and `EXIT_FAILURE`. You should use those instead of 1 and 0, which is hard to remember to do consistently (and in fact, down below you use 1 and -1).